src/modules/adaptors/Frax/CollateralFTokenAdaptor.sol

```diff
@@ -0,0 +1,226 @@
+        bytes memory adaptorData,
+        bytes memory
+    ) public override {
+        // use addCollateral() from fraxlendCore.sol
+    }
+
+    /**
+     * @notice User withdraws are NOT allowed from this position.
+     * TODO: EIN write implementation code for base function
+     */
+    function withdraw(
+        uint256,
+        address,
+        bytes memory,
+        bytes memory
+    ) public pure override {}
+
+    /**
+     * @notice This position is a debt position, and user withdraws are not allowed so
+     *         this position must return 0 for withdrawableFrom.
+     * TODO: EIN write implementation code for base function
+     */
+    function withdrawableFrom(bytes memory, bytes memory) public pure override returns (uint256) {
+        return 0; // TODO: see AaveATokenAdaptor for idea of what to do here. Not sure it applies.
+    }
+
+    /**
+     * @notice Returns the cellar's balance of the collateralAsset position.
+     * @param adaptorData the collateral asset deposited into Fraxlend
+     * NOTE: CRISPY QUESTION - TODO: confirm that this works... EIN - FraxlendCore doesn't have an interface to access the getter. So we'll have to have it here.
+     * TODO: confirm that there is no need to typeCast on the decoded adaptorData when needing address, ERC20 or vice versa.
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        ICollateralFToken fToken = ICollateralFToken(abi.decode(adaptorData, (address))); // TODO: CRISPY QUESTION - could change adaptorData to be collateralToken but then it won't be consistent with rest of fraxlend adaptors.
+        return fToken.userCollateralBalance(msg.sender) + fToken.collateralContract().balanceOf(msg.sender); // reports balance of collateral provided to protocol.
+    }
+
+    /**
+     * @notice Returns collateral asset
+     */
+    function assetOf(bytes memory) public view override returns (ERC20) {
+        IFToken fToken = abi.decode(adaptorData, (IFToken));
+        return ERC20(fToken.collateralContract());
+    }
+
+    /**
+     * @notice This adaptor returns collateral, and not debt.
+     */
+    function isDebt() public pure override returns (bool) {
+        return false;
+    }
+
+    //============================================ Strategist Functions ===========================================
+
+    /**
+     * @notice Allows strategists to add collateral to the respective cellar position on FraxLend, enabling borrowing.
+     * @dev `borrowTokens()` within DebtFTokenAdaptor can increase cellar collateralPosition too. `borrowTokens()` cannot be used without a trusted CollateralFTokenAdaptor position in the same cellar because CollateralFTokenAdaptors track the collateral within a Fraxlend pair. CRISPY QUESTION - See TODO: within DebtFTokenAdaptor regarding this.
+     */
+    function addCollateral(
+        ERC20 _collateralToken,
+        uint256 _collateralToDeposit,
+        IFToken _fToken
```

Comment:
I think adaptordata should contain the fToken address, and the collateral token. Then we can verify this position is setup in the cellar, and that the fToken address is a real fToken address.

Reply:
That makes sense. Sweet, thanks!